### ISSUE TITLE
IDE doesn't honor preprocessor conditionals when including libraries

### ISSUE THREAD

**Issue by tinman89** [arduino/Arduino#1841]
In the Arduino IDE 1.5.5, the default sketch on the Uno compiles to 466 bytes:

```C
//this compiles to 466 bytes
void setup() {
}

void loop() {
}
```

Including Servo.h, it compiles to 1104:

```C
//this compiles to 1104 bytes
#include <Servo.h>

void setup() {
}

void loop() {
}
```

I would therefore expect the following code to compile to 466 bytes, but this is not the case:

```C
//this compiles to 1104 bytes even though Servo.h should not be included
#ifdef SOME_MACRO_THAT_IS_NOT_DEFINED
#include <Servo.h>
#endif

void setup() {
}

void loop() {
}
```

However, commenting out the include statements does prevent the library from being included:

```C
//this compiles to 466 bytes as expected
//#ifdef SOME_MACRO_THAT_IS_NOT_DEFINED
//#include <Servo.h>
//#endif

void setup() {
}

void loop() {
}
```

This caused some headache for me as I tried to figure out why my sketch size was so large despite not including any libraries. This is not the behavior I think anyone would expect from the IDE.

**Comment by ffissore** [comment]
Please checkout this thread [1] on devs mailing list and try one of the linked IDEs. It should fix the problems reported in this issue
[1] https://groups.google.com/a/arduino.cc/forum/#!topic/developers/4X2T3rCxXWM

**Comment by ffissore** [comment]
A new version of the new preprocessor is available. Links available at the bottom of this email
https://groups.google.com/a/arduino.cc/forum/#!msg/developers/4X2T3rCxXWM/YNJl6PZuDucJ

**Comment by ffissore** [comment]
New preprocessor tracked at #2636. Builds for testing it are available

**Comment by ffissore** [comment]
Fixed by #3779
